### PR TITLE
Geant4 sensitive detectors clean-up 6

### PR DESCRIPTION
**Pull request by civanch**
This PR includes clean-ups for muon Geant4 sensitive detectors and additional modifications in the base class, tracker and calo sensitive detectors:

1) TrackInformation object access is implemented in the base class and it is used in Calo, Tracker, Muon derived classes (removed multiple implementations of the same code)
2) Fixed dependences between these packages - removed circular dependence between SimG4CMS/SimMuon and SimG4Core/Application, added few really needed dependencies
3) Unified transformation from Geant4 space coordinates to CMS coordinates in muon and and tracker 
4) Use "#include cstdint" instead of boost
5) In MuonSensitiveDetector there were computations needed only for debug printout, these computations are protected by #ifdef 
6) Use LogVerbatim and LogDebug where possible, removed abort() and cerr, use cms exceptions instead
7) Use const pointers to Geant4 objects
8) clean up Observer interface - make update() methods protected
9) removed unnecessary methods, header, unreasonable comments, old commented lines and add new comments

Correctness of this PR means no change of the results.

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by civanch** [comment]
@fabiocos , this PR now provides identical results versus baseline. Can you , please, check and merge it. It includes bug-fix of circular dependence of SIM packages, the effect of which are unclear to me.

File: SimG4CMS/Muon/src/MuonSensitiveDetector.cc [hunk 1]
```diff
@@ -94,64 +92,57 @@ MuonSensitiveDetector::~MuonSensitiveDetector() {
...
     } else {
-      storeVolumeAndTrack(aStep);
-      return false;
+      thePV = aStep->GetPreStepPoint()->GetPhysicalVolume();
+      theTrackID = aStep->GetTrack()->GetTrackID();
+      theDetUnitId = newDetUnitId;
     }
   }
-  return false;
+  return true;
```
Comment (fabiocos) [inline SimG4CMS/Muon/src/MuonSensitiveDetector.cc]: @civanch for my understanding, what is implied by this change of logic? What was the triggered action, if any, by a "false" value returned?

File: SimG4CMS/Muon/interface/MuonSensitiveDetector.h [hunk 2]
```diff
@@ -36,22 +29,18 @@ class UpdatablePSimHit;
...
-class G4Track;
+class G4Step;
 class G4ProcessTypeEnumerator;
-class G4TrackToParticleID;
 class SimTrackManager;
 
 class MuonSensitiveDetector : 
 public SensitiveTkDetector,
-public Observer<const BeginOfEvent*>,
-public Observer<const EndOfEvent*>
```
Comment (fabiocos) [inline SimG4CMS/Muon/interface/MuonSensitiveDetector.h]: @civanch there is no action in the EndOfEvent, this is the reason to remove this line, am I correct?

File: SimG4CMS/Muon/src/MuonSensitiveDetector.cc [hunk 3]
```diff
@@ -20,72 +19,71 @@
...
 #include "G4SystemOfUnits.hh"
+#include "G4Step.hh"
+#include "G4StepPoint.hh"
+#include "G4Track.hh"
 
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 
 #include <iostream>
 
+//#define DebugLog
```
Comment (fabiocos) [inline SimG4CMS/Muon/src/MuonSensitiveDetector.cc]: @civanch I understand that you keep the revision of the verbosity that we discussed for a further PR, am I correct?

File: SimG4CMS/Muon/src/MuonSensitiveDetector.cc (hunk 1 above)
Comment (civanch) [reply SimG4CMS/Muon/src/MuonSensitiveDetector.cc]: There is no difference to me - the returned bool does not used anywhere. I put "true" in order not make impression to newcomer, that a choice should be done between "true" and "false".

File: SimG4CMS/Muon/interface/MuonSensitiveDetector.h (hunk 2 above)
Comment (civanch) [reply SimG4CMS/Muon/interface/MuonSensitiveDetector.h]: Yes, if methods will not be empty, they may be returned back.

File: SimG4CMS/Muon/src/MuonSensitiveDetector.cc (hunk 3 above)
Comment (civanch) [reply SimG4CMS/Muon/src/MuonSensitiveDetector.cc]: In this code it turned out that a significant part of computation is done only for debugging. It much less work is to include #ifdef, LogDebug itself cannot serving for that.

Comment (civanch) [reply SimG4CMS/Muon/src/MuonSensitiveDetector.cc]: @fabiocos , in this PR I tried improving verbosity using Verbatim stream and LogDebug where possible but in two places seems the best solution is to add #ifdef

File: SimG4CMS/Muon/interface/MuonSensitiveDetector.h (hunk 2 above)
Comment (civanch) [reply SimG4CMS/Muon/interface/MuonSensitiveDetector.h]: @fabiocos , one of the goals of this clean-up is to take out methods/code which doing nothing but is copy/paste from one SD class to another.

File: SimG4CMS/Muon/src/MuonSensitiveDetector.cc (hunk 3 above)
Comment (fabiocos) [reply SimG4CMS/Muon/src/MuonSensitiveDetector.cc]: @civanch I was referring to the standardisation to EDM_ML_DEBUG...

**Comment by fabiocos** [comment]
+1

**Pull request merged**